core/storage/question/gae_models.py

```diff
@@ -357,59 +354,88 @@ def get_question_skill_links_based_on_difficulty_equidistributed_by_skill(
 
         question_skill_link_mapping = {}
 
+        def get_offset(query):
+            """Helper function to get the offset."""
+            question_count = query.count()
+            if question_count > 2 * question_count_per_skill:
+                return utils.get_random_int(
+                    question_count - (question_count_per_skill * 2))
+            return 0
+
         for skill_id in skill_ids:
             query = cls.query(cls.skill_id == skill_id)
 
             equal_questions_query = query.filter(
                 cls.skill_difficulty == difficulty_requested)
+
             # We fetch more questions here in order to try and ensure that the
             # eventual number of returned questions is sufficient to meet the
             # number requested, even after deduplication.
-            new_question_skill_link_models = (
-                equal_questions_query.fetch(question_count_per_skill * 2))
+            new_question_skill_link_models = equal_questions_query.fetch(
+                limit=question_count_per_skill * 2,
+                offset=get_offset(equal_questions_query))
             for model in new_question_skill_link_models:
                 if model.question_id in question_skill_link_mapping:
                     new_question_skill_link_models.remove(model)
 
-            if len(new_question_skill_link_models) < question_count_per_skill:
+            if len(new_question_skill_link_models) >= question_count_per_skill:
+                new_question_skill_link_models = utils.get_random_sample(
+                    new_question_skill_link_models, question_count_per_skill)
+            else:
                 # Fetch QuestionSkillLinkModels with difficulty smaller than
                 # requested difficulty and sort them by decreasing difficulty.
```

Comment:
I think we need to change these comments considering we are not ordering them by skill difficulty anymore

Reply:
Done.